application/common/models/User.php

```diff
@@ -0,0 +1,190 @@
+                $this->validatePassword(),
+                'on' => self::SCENARIO_AUTHENTICATE,
+            ],
+            [
+                ['last_synced_utc', 'last_changed_utc'],
+                'default', 'value' => Utils::now(),
+            ],
+            [
+                ['employee_id', 'first_name', 'last_name', 'display_name', 'username', 'email'],
+                'string',
+            ],
+        ], parent::rules());
+    }
+
+    private function validatePassword(): \Closure
+    {
+        return function ($attributeName) {
+            if (! password_verify($this->$attributeName, $this->password_hash)) {
+                $this->addError($attributeName, 'Incorrect password.');
+            }
+        };
+    }
+
+    public function behaviors(): array
+    {
+        return [
+            'changeTracker' => [
+                'class' => AttributeBehavior::className(),
+                'attributes' => [
+                    ActiveRecord::EVENT_BEFORE_INSERT => 'last_changed_utc',
+                    ActiveRecord::EVENT_BEFORE_UPDATE => 'last_changed_utc',
+                ],
+                'value' => Utils::now(),
+                'skipUpdateOnClean' => true, // only update the column if the model is dirty
+            ],
+            'updateTracker' => [
+                'class' => AttributeBehavior::className(),
+                'attributes' => [
+                    ActiveRecord::EVENT_BEFORE_INSERT => 'last_synced_utc',
+                    ActiveRecord::EVENT_BEFORE_UPDATE => 'last_synced_utc',
+                ],
+                'value' => Utils::now(),
+                'skipUpdateOnClean' => false, // always update the column regardless of dirtiness
+            ],
+        ];
+    }
+
+    /**
+     * @return array of fields that should be included in responses.
+     */
+    public function fields(): array
+    {
+        return [
+            'employee_id',
+            'first_name',
+            'last_name',
+            'display_name',
+            'username',
+            'email',
+            'active',
+            'locked',
+            'last_changed_utc',
+            'last_synced_utc',
+        ];
+    }
+
+    public function save($runValidation = true, $attributeNames = null)
+    {
+        if ($this->scenario === self::SCENARIO_UPDATE_PASSWORD) {
+            return $this->savePassword($runValidation, $attributeNames);
+        }
+
+        return parent::save($runValidation, $attributeNames);
+    }
+
+    private function savePassword($runValidation = true, $attributeNames = null): bool
+    {
+        $transaction = ActiveRecord::getDb()->beginTransaction();
+
+        try {
+            if ($this->hasPasswordAlready()) {
+                if (! $this->saveHistory()) {
+                    return false;
+                }
+            }
+
+            $this->password_hash = password_hash($this->password, PASSWORD_DEFAULT);
+
+            if (! parent::save($runValidation, $attributeNames)) {
```

Comment:
I think my concern is that, if this `savePassword(...)` method was called in a scenario other than `UPDATE_PASSWORD`, it would save more than the password. Based on it's name, I would think we'd want to remove the `$attributeNames` parameter to this `savePassword` function and here in the call to `parent::save(...)` specifically instruct it to only save the `password` (and/or `password_hash`... I'm not sure which). It's one of those situations where it feels like we're saying "oh, some other piece of code will enforce the proper constraints for this function" rather than having the function only do what it's name claims it will do.

Thoughts?